pkg/tests/mappings/mappings_external_repo_test.bzl

```diff
@@ -0,0 +1,183 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for file mapping routines in pkg/mappings.bzl"""
+
+load("@bazel_skylib//lib:new_sets.bzl", "sets")
+load("@bazel_skylib//lib:unittest.bzl", "analysistest", "asserts", "unittest")
+load("//:providers.bzl", "PackageFilegroupInfo", "PackageFilesInfo")
+load("//:mappings.bzl", "pkg_files", "strip_prefix")
+
+##########
+# Helpers
+##########
+
+def _flatten(list_of_lists):
+    """Transform a list of lists into a single list, preserving relative order."""
+    return [item for sublist in list_of_lists for item in sublist]
+
+##########
+# pkg_files tests
+##########
+
+def _pkg_files_contents_test_impl(ctx):
+    env = analysistest.begin(ctx)
+    target_under_test = analysistest.target_under_test(env)
+
+    expected_dests = {e: None for e in ctx.attr.expected_dests}
+    n_found = 0
+    for got in target_under_test[PackageFilesInfo].dest_src_map.keys():
+        asserts.true(
+            got in expected_dests,
+            "got <%s> not in expected set: %s" % (got, ctx.attr.expected_dests))
+        n_found += 1
+    asserts.equals(env, len(expected_dests), n_found)
+
+
+    # Simple equality checks for the others, if specified
+    if ctx.attr.expected_attributes:
+        asserts.equals(
+            env,
+            json.decode(ctx.attr.expected_attributes),
+            target_under_test[PackageFilesInfo].attributes,
+            "pkg_files attributes do not match expectations",
+        )
+
+    # TODO(nacl): verify DefaultInfo propagation
+
+    return analysistest.end(env)
+
+pkg_files_contents_test = analysistest.make(
+    _pkg_files_contents_test_impl,
+    attrs = {
+        # Other attributes can be tested here, but the most important one is the
+        # destinations.
+        "expected_dests": attr.string_list(
+            mandatory = True,
+        ),
+        "expected_attributes": attr.string(),
+    },
+)
```

Comment:
Consider putting these functions in a local .bzl file sourced by both the local and external repository tests.